### PR TITLE
Add PauliList class

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Add `PauliList` to replace `PauliTable` and `StabilizerTable`

### Details and comments

This class functions as an optimized list of the new `Pauli` class (including phase).

- [x] tests
- [x] Improve documentation

### CODE REVIEW

File: qiskit/quantum_info/operators/symplectic/pauli_list.py [hunk 1]
```diff
@@ -0,0 +1,1057 @@
...
+            base_phase = np.insert(self._phase, ind, value._phase)
+            return PauliList(BasePauli(base_z, base_x, base_phase))
+
+        # Column insertion
+        if ind > self.num_qubits:
+            raise QiskitError(
+                "Index {} is greater than number of qubits"
+                " in the PauliList ({})".format(ind, self.num_qubits)
+            )
+        # TODO: Fix phase
```
Comment (ShellyGarion) [inline qiskit/quantum_info/operators/symplectic/pauli_list.py:417]: is this TODO still needed?

File: qiskit/quantum_info/operators/symplectic/pauli_list.py [hunk 2]
```diff
@@ -0,0 +1,1057 @@
...
+        """
+        from qiskit.circuit import Instruction, QuantumCircuit
+        from qiskit.quantum_info.operators.symplectic.clifford import Clifford
+
+        if qargs is None:
+            qargs = getattr(other, "qargs", None)
+
+        # Convert quantum circuits to Cliffords
+        if isinstance(other, Clifford):
+            other = other.to_circuit()
```
Comment (ShellyGarion) [inline qiskit/quantum_info/operators/symplectic/pauli_list.py:886]: can you maybe do it directly without decomposing the cliffolrd? like its done in #6297

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py [hunk 3]
```diff
@@ -0,0 +1,1470 @@
...
+            target = StabilizerTable.from_labels(["+II", "-XZ"])
+            value = PauliList(target)
+            self.assertEqual(value, target)
+
+        with self.subTest(msg="PauliTable no copy"):
+            target = StabilizerTable.from_labels(["+YY", "-XZ", "XI"])
+            value = PauliList(target)
+            value[0] = "II"
+            self.assertEqual(value, target)
+
```
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:178]: In the init tests - maybe worth checking more phases with `i` and `-i`

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py [hunk 4]
```diff
@@ -0,0 +1,1470 @@
...
+    def test_add_methods(self):
+        """Test __add__ method."""
+        labels1 = ["XXI", "IXX"]
+        labels2 = ["XXI", "ZZI", "ZYZ"]
+        pauli1 = PauliList(labels1)
+        pauli2 = PauliList(labels2)
+        target = PauliList(labels1 + labels2)
+        self.assertEqual(target, pauli1 + pauli2)
+
+    def test_add_qargs(self):
```
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:284]: can test also pseudo-random pauli's

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py [hunk 5]
```diff
@@ -0,0 +1,1470 @@
...
+            pauli[1:3] = target
+            self.assertEqual(pauli[1:3], target)
+
+
+class TestPauliListLabels(QiskitTestCase):
+    """Tests PauliList label representation conversions."""
+
+    def test_from_labels_1q(self):
+        """Test 1-qubit from_labels method."""
+        labels = ["I", "Z", "Z", "X", "Y"]
```
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:368]: Z appears twice

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py [hunk 6]
```diff
@@ -0,0 +1,1470 @@
...
+        value = PauliList(labels)
+        self.assertEqual(target, value)
+
+    def test_to_labels_1q(self):
+        """Test 1-qubit to_labels method."""
+        pauli = PauliList.from_symplectic(
+            np.array([[False], [True], [True], [False], [True]]),
+            np.array([[False], [False], [False], [True], [True]]),
+        )
+        target = ["I", "Z", "Z", "X", "Y"]
```
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:409]: Z appears twice

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py [hunk 7]
```diff
@@ -0,0 +1,1470 @@
...
+        """Test 1-qubit to_labels method w/ array=True."""
+        pauli = PauliList.from_symplectic(
+            np.array([[False], [True], [True], [False], [True]]),
+            np.array([[False], [False], [False], [True], [True]]),
+        )
+        target = np.array(["I", "Z", "Z", "X", "Y"])
+        value = pauli.to_labels(array=True)
+        self.assertTrue(np.all(value == target))
+
+    def test_labels_round_trip(self):
```
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:423]: can try more phases as well as pseudo-random paulis

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py [hunk 8]
```diff
@@ -0,0 +1,1470 @@
...
+        """Test PauliList sparse matrix_iter method."""
+        labels = ["III", "IXI", "IYY", "YIZ", "XYZ", "III"]
+        pauli = PauliList(labels)
+        for idx, i in enumerate(pauli.matrix_iter(sparse=True)):
+            self.assertTrue(isinstance(i, csr_matrix))
+            self.assertTrue(np.all(i.toarray() == pauli_mat(labels[idx])))
+
+
+@ddt
+class TestPauliListOperator(QiskitTestCase):
```
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:559]: in operators tests can test more phases and also pseudo-random paulis

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py [hunk 9]
```diff
@@ -0,0 +1,1470 @@
...
+            self.assertEqual(value, target)
+
+        with self.subTest(msg="dot 3-qubit qargs=[1, 0, 2]"):
+            target = PauliList(["XZY", "IYZ"])
+            value = pauli1.dot(pauli2, qargs=[1, 0, 2])
+            self.assertEqual(value, target)
+
+
+@ddt
+class TestPauliListMethods(QiskitTestCase):
```
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:784]: in sorting tests can test more phases

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py [hunk 10]
```diff
@@ -0,0 +1,1470 @@
...
+            self.assertEqual(target, value)
+
+        with self.subTest(msg="10 qubit"):
+            labels = [10 * "X", 10 * "I", 10 * "X"]
+            unique = [10 * "X", 10 * "I"]
+            target = PauliList(unique)
+            value = PauliList(labels).unique()
+            self.assertEqual(target, value)
+
+    def test_delete(self):
```
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:1040]: in delete and insert tests - maybe add phases?

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py [hunk 11]
```diff
@@ -0,0 +1,1470 @@
...
+                target0 = PauliList(["X" + i[0], "Y" + i[1], "Z" + i[2]])
+                target1 = PauliList([i[0] + "X", i[1] + "Y", i[2] + "Z"])
+
+                with self.subTest(msg="multiple-column multiple-vals from PauliList"):
+                    value = pauli.insert(0, PauliList(i), qubit=True)
+                    self.assertEqual(value, target0)
+                    value = pauli.insert(1, PauliList(i), qubit=True)
+                    self.assertEqual(value, target1)
+
+    def test_commutes(self):
```
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:1178]: in commute / anti-commute tests - maybe add phases?

File: qiskit/quantum_info/operators/symplectic/pauli_list.py (hunk 1 above)
Comment (ikkoham) [reply qiskit/quantum_info/operators/symplectic/pauli_list.py:417]: Thank you for your pointing out. I forget fixing the phase.

File: qiskit/quantum_info/operators/symplectic/pauli_list.py (hunk 2 above)
Comment (ikkoham) [reply qiskit/quantum_info/operators/symplectic/pauli_list.py:886]: Thanks. I've moved the _evolve_clifford to the base class.

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py (hunk 3 above)
Comment (ikkoham) [reply test/python/quantum_info/operators/symplectic/test_pauli_list.py:178]: Thanks. I added the tests.

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py (hunk 4 above)
Comment (ikkoham) [reply test/python/quantum_info/operators/symplectic/test_pauli_list.py:284]: Thanks. Why do we need the test for the random Paulis here?

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py (hunk 5 above)
Comment (ikkoham) [reply test/python/quantum_info/operators/symplectic/test_pauli_list.py:368]: This is as intended. We also see that the same Pauli is not omitted.

**Comment by ikkoham** [comment]
@ShellyGarion Thank you very much for your review. The missing tests about the phase were fatal, and I forgot the TODO. I am thankful for your review helped me realize them.

I have fixed them in the PR: https://github.com/chriseclectic/qiskit-terra/pull/24

File: qiskit/quantum_info/operators/symplectic/pauli_list.py [hunk 12]
```diff
@@ -0,0 +1,1064 @@
...
+        return self._truncated_str(False)
+
+    def _truncated_str(self, show_class):
+        stop = self._num_paulis
+        if self.__truncate__:
+            max_paulis = self.__truncate__ // self.num_qubits
+            if self._num_paulis > max_paulis:
+                stop = max_paulis
+        labels = [str(self[i]) for i in range(stop)]
+        prefix = "PauliList(" if show_class else ""
```
Comment (t-imamichi) [inline qiskit/quantum_info/operators/symplectic/pauli_list.py:192]: According to python docs, a form "<...>" is suggested as follows. How about changing prefix to "<PauliList("?
> Called by the repr() built-in function to compute the “official” string representation of an object. If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment). If this is not possible, a string of the form <...some useful description...> should be returned.
https://docs.python.org/3/reference/datamodel.html#object.__repr__

Comment (ikkoham) [reply qiskit/quantum_info/operators/symplectic/pauli_list.py:192]: Thank you for your review. `PauliList.__repr__` is evaluatable unless it is truncated. So, this is valid Python expression.

**Review by ShellyGarion — APPROVED** [review]

**Review by t-imamichi — DISMISSED** [review]

**Review by mtreinish — APPROVED** [review]
LGTM

**Review by mtreinish — CHANGES_REQUESTED** [review]
Actually I missed it initially, this needs a release note documenting the new class. After that's added this is ready to go

**Comment by ikkoham** [comment]
@mtreinish Thank you. I made a PR to add the release note. https://github.com/chriseclectic/qiskit-terra/pull/25

@chriseclectic Could you merge this PR if it's OK.

**Review by mtreinish — APPROVED** [review]
LGTM, I just merged @ikkoham's commit with the release notes from https://github.com/chriseclectic/qiskit-terra/pull/25

**Comment by mtreinish** [comment]
It looks like there is a conflict here now: https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=29265&view=logs&j=377f2c8d-8b6f-583f-e152-2ff025e2543a&t=28b3ff0d-9e76-5e94-89e9-eed7321a383e&l=12999 probably after #6442 merged

**Comment by ikkoham** [comment]
OK. I'll fix this soon.

**Comment by ikkoham** [comment]
Thank you. I think this PR fixes the error on CI. (I've tested on my local and it passed.)
https://github.com/chriseclectic/qiskit-terra/pull/26

**Review by mtreinish — APPROVED** [review]

**Pull request merged**